### PR TITLE
Kivy: use the same password for all wallets

### PR DESCRIPTION
**Pull request by ecdsa**
When a wallet password is updated, check the password for all
wallets in directory. If that check passes, update password for
all wallets. If it fails, update only the current wallet.

### CODE REVIEW

**Comment by ecdsa** [comment]
note: once this is enabled, we should let users reset their PIN with their password

File: electrum/gui/kivy/main_window.py [hunk 1]
```diff
@@ -649,6 +649,10 @@ def load_wallet_by_name(self, path):
             return
         if self.wallet and self.wallet.storage.path == path:
             return
+        if self.password and self.electrum_config.get('use_single_password'):
+            storage = WalletStorage(path)
+            storage.check_password(self.password)
```
Comment (SomberNight) [inline electrum/gui/kivy/main_window.py:654]: what if wallet is only keystore-encrypted?

File: electrum/gui/kivy/main_window.py [hunk 2]
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
...
+    def check_password_for_directory(self, old_password, new_password, update=False):
+        dirname = os.path.dirname(self.electrum_config.get_wallet_path())
+        self.logger.info(f'update pw {dirname}')
+        failed = []
+        success = []
+        for filename in os.listdir(dirname):
+            path = os.path.join(dirname, filename)
+            basename = os.path.basename(path)
+            storage = WalletStorage(path)
+            if not storage.is_encrypted():
```
Comment (SomberNight) [inline electrum/gui/kivy/main_window.py:1230]: what is wallet is already keystore-encrypted (but not storage-encrypted)?

File: electrum/gui/kivy/main_window.py [hunk 3]
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
...
+                # let's encrypt it
+                db = WalletDB(storage.read(), manual_upgrades=False)
+                wallet = Wallet(db, storage, config=self.electrum_config)
```
Comment (SomberNight) [inline electrum/gui/kivy/main_window.py:1233]: maybe these lines too should be guarded with `if update`?

File: electrum/gui/kivy/main_window.py [hunk 4]
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
...
+                success.append(basename)
+                if update:
+                    wallet.update_password(None, new_password)
```
Comment (SomberNight) [inline electrum/gui/kivy/main_window.py:1236]: I guess `InvalidPassword` is not handled as you assume it cannot be raised due to `if not storage.is_encrypted()` - but again, wallet might be keystore-encrypted

File: electrum/gui/kivy/main_window.py [hunk 5]
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
         if pin != self.electrum_config.get('pin_code'):
             raise InvalidPassword
 
+    def check_password_for_directory(self, old_password, new_password, update=False):
```
Comment (SomberNight) [inline electrum/gui/kivy/main_window.py:1221]: Could you either add a docstring or change the API?
E.g. 
```
"""Checks password against all wallets and returns (list of matches, list of mismatches).
If update is True, it also changes the wallet passwords to new_password.
"""
```
Alternatively, the API might be clearer if "update" and "new_password" are merged together. Simplest would be to set `new_password` to `None` to correspond to `update==False`, however `None` is a valid value for `new_password` I guess, so we could introduce a sentinel e.g.
```
SENTINEL=object()
def check_password_for_directory(self, old_password, new_password=SENTINEL): ...
```

File: electrum/gui/kivy/main_window.py [hunk 6]
```diff
@@ -1213,11 +1218,60 @@ def check_pin_code(self, pin):
...
         def on_success(old_password, new_password):
-            self.wallet.update_password(old_password, new_password)
+            # called if old_password works on self.wallet
             self.password = new_password
-            self.show_info(_("Your password was updated"))
+            success, failed = self.check_password_for_directory(old_password, new_password)
+            if not failed:
+                self.check_password_for_directory(old_password, new_password, update=True)
+                self.electrum_config.set_key('use_single_password', True)
+                msg = _(f"Password updated successfully for {len(success)} wallets")
```
Comment (SomberNight) [inline electrum/gui/kivy/main_window.py:1269]: this does not work, see https://github.com/spesmilo/electrum/blob/78513affe536ba7d8150dc2ee044e3258097317f/electrum/i18n.py#L33-L35

File: electrum/gui/kivy/main_window.py (hunk 5 above)
Comment (ecdsa) [reply electrum/gui/kivy/main_window.py:1211]: I agree, it is better to change the API. `None` is not a valid password for Kivy, so I guess it is fine to use it as long as this method stays in the GUI

File: electrum/gui/kivy/main_window.py (hunk 3 above)
Comment (ecdsa) [reply electrum/gui/kivy/main_window.py:1223]: ack

**Comment by ecdsa** [comment]
note: if a user does not know the unique password, they can still create a new wallet, and this will result in a situation where the unique password is disabled. This could be annoying in case of an unauthorized third party accessing the phone.

File: electrum/gui/kivy/main_window.py (hunk 1 above)
Comment (ecdsa) [reply electrum/gui/kivy/main_window.py:660]: This case is handled in `check_password_for_directory`. `self._use_single_password` is set only if the password was checked against all wallets and keystores. Here check_password is called for its side effect, to decrypt the storage. I will add a comment to clarify.

File: electrum/gui/kivy/main_window.py (hunk 2 above)
Comment (ecdsa) [reply electrum/gui/kivy/main_window.py]: this is now handled.

File: electrum/gui/kivy/main_window.py (hunk 4 above)
Comment (ecdsa) [reply electrum/gui/kivy/main_window.py]: same as above

**Comment by ecdsa** [comment]
Rebased and squashed to have a single commit. Wallet creation is now disabled if the user does not know their password, so that password unification is irreversible.

An edge case remains: if the first wallet to be open is unencrypted and watching-only, no password is asked and `use_single_password` will be set to False

**Pull request merged**